cluster-autoscaler/core/scaledown/eligibility/eligibility.go

```diff
@@ -0,0 +1,181 @@
+	// GetScaleDownGpuUtilizationThreshold returns ScaleDownGpuUtilizationThreshold value that should be used for a given NodeGroup.
+	GetScaleDownGpuUtilizationThreshold(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (float64, error)
+}
+
+// NewChecker creates a new Checker object.
+func NewChecker(thresholdGetter utilizationThresholdGetter) *Checker {
+	return &Checker{
+		thresholdGetter: thresholdGetter,
+	}
+}
+
+// FilterOutUnremovable accepts a list of nodes that are candidates for
+// scale down and filters out nodes that cannot be removed, along with node
+// utilization info.
+// TODO(x13n): Node utilization could actually be calculated independently for
+// all nodes and just used here. Next refactor...
+func (c *Checker) FilterOutUnremovable(context *context.AutoscalingContext, scaleDownCandidates []*apiv1.Node, timestamp time.Time, unremovableNodes *unremovable.Nodes) ([]string, map[string]utilization.Info) {
+	unremovableNodes.Update(context.ClusterSnapshot.NodeInfos(), timestamp)
+
+	skipped := 0
+	utilizationMap := make(map[string]utilization.Info)
+	currentlyUnneededNodeNames := make([]string, 0, len(scaleDownCandidates))
+	utilLogsQuota := klogx.NewLoggingQuota(20)
+
+	for _, node := range scaleDownCandidates {
+		nodeInfo, err := context.ClusterSnapshot.NodeInfos().Get(node.Name)
+		if err != nil {
+			klog.Errorf("Can't retrieve scale-down candidate %s from snapshot, err: %v", node.Name, err)
+			unremovableNodes.AddReason(node, simulator.UnexpectedError)
+			continue
+		}
+
+		// Skip nodes that were recently checked.
+		if unremovableNodes.IsRecent(node.Name) {
+			unremovableNodes.AddReason(node, simulator.RecentlyUnremovable)
+			skipped++
+			continue
+		}
+
+		reason, utilInfo := c.unremovableReasonAndNodeUtilization(context, timestamp, nodeInfo, utilLogsQuota)
+		if utilInfo != nil {
+			utilizationMap[node.Name] = *utilInfo
+		}
+		if reason != simulator.NoReason {
+			unremovableNodes.AddReason(node, reason)
+			continue
+		}
+
+		currentlyUnneededNodeNames = append(currentlyUnneededNodeNames, node.Name)
+	}
+
+	klogx.V(4).Over(utilLogsQuota).Infof("Skipped logging utilization for %d other nodes", -utilLogsQuota.Left())
+	if skipped > 0 {
+		klog.V(1).Infof("Scale-down calculation: ignoring %v nodes unremovable in the last %v", skipped, context.AutoscalingOptions.UnremovableNodeRecheckTimeout)
+	}
+	return currentlyUnneededNodeNames, utilizationMap
+}
+
+func (c *Checker) unremovableReasonAndNodeUtilization(context *context.AutoscalingContext, timestamp time.Time, nodeInfo *schedulerframework.NodeInfo, utilLogsQuota *klogx.Quota) (simulator.UnremovableReason, *utilization.Info) {
+	node := nodeInfo.Node()
+
+	// Skip nodes marked to be deleted, if they were marked recently.
+	// Old-time marked nodes are again eligible for deletion - something went wrong with them
+	// and they have not been deleted.
```

Comment:
nit: I'd get rid of this comment. The code is self-explanatory and the comment just leaks implementation details of IsNodeBeingDeleted().